<h1><font face="fantasy, serif" color=#533563 style="font-size:40px"><center>MTR 3420 Dynamics Interactive Lab Notebook</center></font></h1>

<h1><font face="fantasy, serif" color=#533563 style="font-size:40px"><center>Data Notebook</center></font></h1>

<h1><font face="fantasy, serif" color=#533563 style="font-size:40px"><center>Justin Richling - 08/20/2018</center></font></h1>

<h1><font face="fantasy, serif" color=#533563 style="font-size:23px"><center>This notebook will allow many learning opportunities:</center><br/></font><font face="fantasy, serif" color=#533563 style="font-size:18px"><ul>
  <li><font>Reading Data from NetCDF file</font></li><br>
  <li>Working with Multidimensional Arrays</li><br>
  <li>Creating Functions</li><br>
  <li>Running Calculations</li><br>
  <li>Plotting Maps</li><br>
  <li>Using Working Notebook - chance to use HTML and Tex formatting</li><br>
  <li>Learning Python and common programming language terminology</li><br>
  <li>Algorithms - Learning how to think rationally</li>
</ul></font></h1>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h2><font face="fantasy, serif" color=#533563 style="font-size:40px"><center>In Python we can import external libraries that we can download which will help us in numerous ways</center><br></font><font face="fantasy, serif" color=#533563 style="font-size:20px">One of the most useful aspects of Python is being able to import external libraries (once you have them downloaded) to help with such things like mathematical calcs, plotting, working with arrays and specific types of data files (NetCDF for example).<br><br>We can also use them for grabbing things from the internet or the operating system of our computer like time or command line operations.  </font></h2>

In [ ]:
# matplotlib is the powerful plotting tool that allows us to make customizable plots
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker

# numpy is Python's math library and helps us with calculations. Trig functions, constants and 
# symbols are accessable.
import numpy as np

# xarray will allow us so many cool and useful features when working with multidimensional data arrays
import xarray as xr

# build our backgroud map and plotting control
from mpl_toolkits.basemap import Basemap

# Locally set the figure size
plt.rcParams['figure.figsize'] = (20, 20)

#
import scipy.ndimage as ndimage

# for better control of the ticks on the x and y-axis
from matplotlib import ticker

#import pandas as pd 
# It wouldn't be a bad idea to save our calulations in a new .nc file so we don't have to run calculation
# everytime we open out notebook, we can just call in the data by opening the new .nc file.
from netCDF4 import Dataset

# If we want to import images the Image libary will be what we want 
from IPython.display import Image, display

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h2><h2><font face="fantasy, serif" color=#533563 style="font-size:40px"><center>Now it's time to pull some data from our NetCDF file!</center><br></font><font face="fantasy, serif" color=#533563 style="font-size:20px">Setting the file path as a variable <font face="fantasy, serif" color=#000000 style="font-size:23px"><i>datafile</i></font>. Find out where this file lives on your computer. We can use the command line or through finder windows on the desktop.<br><br>Then calling xr.open() will read all the data from the NetCDF file and we can print it out to examine what's going on.</font></h2>

<h2><font face="fantasy, serif" color=#533563 style="font-size:20px">What we will be doing time and time again is assigning data, numbers, strings, etc to <font face="fantasy, serif" color=#000000 style="font-size:23px"><i>variables</i></font>. This allows us to keep accessing the data once the data file is closed. Most importantly it will allow us to be free to probe the information stored in the variable over and over again in our <font face="fantasy, serif" color=#000000 style="font-size:23px"><i>for-loops</i></font>.</font></h2>

In [4]:
# Try to find the location of the data file and set the path to the variable dataFile
dataFile = '/Users/chowdahead/Desktop/Dynamics/hgt.2016.nc'

#Open the dataset, set it to a vriable and print out metadeta
ds = xr.open_dataset(dataFile)

# Let's take a look at the metadata
print ds

<xarray.Dataset>
Dimensions:  (lat: 73, level: 17, lon: 144, time: 1464)
Coordinates:
  * level    (level) float32 1000.0 925.0 850.0 700.0 600.0 500.0 400.0 ...
  * lat      (lat) float32 90.0 87.5 85.0 82.5 80.0 77.5 75.0 72.5 70.0 67.5 ...
  * lon      (lon) float32 0.0 2.5 5.0 7.5 10.0 12.5 15.0 17.5 20.0 22.5 ...
  * time     (time) datetime64[ns] 2016-01-01 2016-01-01T06:00:00 ...
Data variables:
    hgt      (time, level, lat, lon) float32 ...
Attributes:
    Conventions:    COARDS
    title:          4x daily NMC reanalysis (2014)
    history:        created 2013/12 by Hoop (netCDF2.3)
    description:    Data is from NMC initialized reanalysis\n(4x/day).  It co...
    platform:       Model
    References:     http://www.esrl.noaa.gov/psd/data/gridded/data.ncep.reana...
    dataset_title:  NCEP-NCAR Reanalysis 1


In [5]:
ds.keys()

/Users/chowdahead/anaconda/lib/python2.7/_abcoll.py:410: FutureWarning: iteration over an xarray.Dataset will change in xarray v0.11 to only include data variables, not coordinates. Iterate over the Dataset.variables property instead to preserve existing behavior in a forwards compatible manner.
  return list(self)
/Users/chowdahead/anaconda/lib/python2.7/_abcoll.py:410: FutureWarning: calling len() on an xarray.Dataset will change in xarray v0.11 to only include data variables, not coordinates. Call len() on the Dataset.variables property instead, like ``len(ds.variables)``, to preserve existing behavior in a forwards compatible manner.
  return list(self)


[u'level', u'lat', u'lon', u'time', u'hgt']

In [6]:
ds['lon']

<xarray.DataArray 'lon' (lon: 144)>
array([   0. ,    2.5,    5. ,    7.5,   10. ,   12.5,   15. ,   17.5,   20. ,
         22.5,   25. ,   27.5,   30. ,   32.5,   35. ,   37.5,   40. ,   42.5,
         45. ,   47.5,   50. ,   52.5,   55. ,   57.5,   60. ,   62.5,   65. ,
         67.5,   70. ,   72.5,   75. ,   77.5,   80. ,   82.5,   85. ,   87.5,
         90. ,   92.5,   95. ,   97.5,  100. ,  102.5,  105. ,  107.5,  110. ,
        112.5,  115. ,  117.5,  120. ,  122.5,  125. ,  127.5,  130. ,  132.5,
        135. ,  137.5,  140. ,  142.5,  145. ,  147.5,  150. ,  152.5,  155. ,
        157.5,  160. ,  162.5,  165. ,  167.5,  170. ,  172.5,  175. ,  177.5,
        180. ,  182.5,  185. ,  187.5,  190. ,  192.5,  195. ,  197.5,  200. ,
        202.5,  205. ,  207.5,  210. ,  212.5,  215. ,  217.5,  220. ,  222.5,
        225. ,  227.5,  230. ,  232.5,  235. ,  237.5,  240. ,  242.5,  245. ,
        247.5,  250. ,  252.5,  255. ,  257.5,  260. ,  262.5,  265. ,  267.5,
        270. ,  

In [7]:
ds['lon'][140]

# actual_range:   [   0.   357.5]  
# means (0,2.5,...,355,357.5)

<xarray.DataArray 'lon' ()>
array(350.0, dtype=float32)
Coordinates:
    lon      float32 350.0
Attributes:
    units:          degrees_east
    long_name:      Longitude
    actual_range:   [   0.   357.5]
    standard_name:  longitude
    axis:           X

<h2><font face="fantasy, serif" color=#533563 style="font-size:23px">We are going to grab all the values for each variable and assign them to a new arrays: Lat, Lon, Time, and Level:</font></h2>

In [8]:
# The .values command is fetching just the numerical data

Lat = ds['lat'].values
Lon = ds['lon'].values
Time = ds['time'].values
Level = ds['level'].values

print "Data type of our new arrays: ",type(Lon)
print "Shape of Lat array: ",Lat.shape[0]
print "Shape of Lon array: ",Lon.shape[0]
print "Shape of Time array: ",Time.shape[0]
print "Shape of Level array: ",Level.shape[0]
print "Number of dimensions for Level array: "+str(len(Level.shape))+", meaning 1 dimension of length 17"

Data type of our new arrays:  <type 'numpy.ndarray'>
Shape of Lat array:  73
Shape of Lon array:  144
Shape of Time array:  1464
Shape of Level array:  17
Number of dimensions for Level array: 1, meaning 1 dimension of length 17


<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

In [9]:
# Lets find the index number in Level array for say 500mb
mylevel = input("Which mb level: ")
mylev = np.where(Level == mylevel)[0][0]
print "Pressure level index of input: ",mylev

Which mb level: 500
Pressure level index of input:  5


<h2><font face="fantasy, serif" color=#533563 style="font-size:23px">Now we need to find which index number Denver's lat and lon are knowing that Denver is approximately 40 N and -105 W or 255 E.<br/><br/> We can use the indexing to find Denver's lat and lon in the same way we did before with the pressure level: </font></h2>

In [15]:
mylat = input("Which latitude? (-90 to 90) ")
Denlat = np.where(Lat == mylat)[0][0]
print "Index number for Denver lat: ",Denlat,"\n" # The \n just means print new line

mylon = input("Which latitude? (0 to 357.5) ") # Think about what is the range of the lons in this data set?
Denlon = np.where(Lon == mylon)[0][0]
print "Index number for Denver lon: ",Denlon

Which latitude? (-90 to 90) 40
Index number for Denver lat:  20 

255
Index number for Denver lon:  102


<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">So Lat[20] = 40 deg north = Denver's Lat <br><br>And Lon[102] = 255 deg east = Denver's Lon.</font><br/ ></h2>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">Finally we want to know the index number for Feb. 2nd, 12:00Z is in the $Time$ array:</font></h2>

In [10]:
january=31*4 #31 days in January, multiplied by 4 for 00, 06, 12, and 18Z each day
groundhog=4+3 #4 time steps on the 1st plus 0 z on the 2nd, 6 z on the 2nd, 12 z on the second 
mytime=january+groundhog-1 # Reason for the minus one??
print "Day of year index number: ",mytime,"\n"

# mytime is the variable for the index number.

Day of year index number:  130 



In [11]:
# We can pull various things out of the metadata like parts of the date and store them in 
# variables for later if we want to help automate things we print out

print "What the raw datetime data looks like: ",Time[mytime],"\n"
print "What the raw datetime data type is: ",type(Time[mytime])

What the raw datetime data looks like:  2016-02-02T12:00:00.000000000 

What the raw datetime data type is:  <type 'numpy.datetime64'>


In [12]:
# We can clean this up a bit with a little trickery...
date1 = str(Time[mytime])[0:10] # Remember we can take pieces of the string
date2 = str(Time[mytime])[-18:-10] # some more pieces...
date = date1+" "+date2 # Finally combine the pieces to make a cleaner date string
# What is printing date going to show?
#print date

hour = date[11:16]
year = date[0:4]
print "Modified string date: ",date
print "Year: ",year
print "Modified hour of day: "+str(hour)+"Z"

Modified string date:  2016-02-02 12:00:00
Year:  2016
Modified hour of day: 12:00Z


<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">Now we can populate a 4d array of time, level, lat, and lon. Each entry in the new 4d array will have a height value in meters:</font></h2>

In [13]:
# hgt(time, level, lat, lon)
height = ds['hgt'][:,:,:,:].values
list(np.float_(height))
print "Shape of the height array: ",height.shape
print type(height[0][0][0][0])

Shape of the height array:  (1464, 17, 73, 144)
<type 'numpy.float32'>


<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">One cool feature of accessing metadata is getting variables to save as variables too. Here we are pulling the description of the name of the variable. For height it will print geopotential heights:</font></h2>

In [10]:
# This will work, any guess why this worked and the previous attempt failed?
# height array stored all the actual values of ds["hgt"], but not the metadata like description (.var_descr)

hgt_descr = str(ds['hgt'].var_desc)
print "What 'hgt' variable is: ",hgt_descr

What 'hgt' variable is:  Geopotential height


<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">Since we've loaded all our data we can close the file so it's not running in the background:</font></h2>

In [16]:
ds.close()

In [ ]:
# Uncomment these if you just want to assign to variables and move on quickly...

#Denlot = 20
#Denlat = 102
#mytime = 130

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

In [ ]:
hgt130_500mb=height[mytime][mylev][:][:]

print type(hgt130_500mb)
print hgt130_500mb.shape


In [ ]:
lat_0 = 20
lat_1 = 60
lon_0 = 225
lon_1 = 300 

#lat_0 = -90
#lat_1 = 90
#lon_0 = 0
#lon_1 = 357.5 

# Initiate a new figure so we can draw on it
fig=plt.figure()

# We need to figure out what set of numbers to plot with and we can pull those values right from 
# the data with .max/min(). 
# np.arange just guarantess that we have consistant plotting and labeling. 
# np.arange(start,end,interval)
hgt130_Surf=height[mytime][0][:][:]
hgt130_Surf_levels = np.arange(0,400,40) 

hgt130_500mb_levels = np.arange(hgt130_500mb.min(),hgt130_500mb.max()+60,60) 

#Smooth the height field 
hgt130_500mb_smooth = ndimage.gaussian_filter(hgt130_500mb, sigma=1, order=0)
hgt130_Surf_smooth = ndimage.gaussian_filter(hgt130_Surf, sigma=1, order=0)

# Setting up our map through the basemap library and calling the function Basemap() and setting 
# it to a variable m (short for map)
m=Basemap(projection='mill',lat_ts=10,llcrnrlon=lon_0, \
      urcrnrlon=lon_1,llcrnrlat=lat_0,urcrnrlat=lat_1, \
      resolution='l')


x, y = m(*np.meshgrid(Lon,Lat))

lons = [262.5]
lats = [35.5]

x1,y1 = m(lons, lats)
labels = ['L']
for label, xpt, ypt in zip(labels, x1, y1):   #fontname='Times New Roman'
    plt.text(xpt, ypt, label,fontsize=70,color='r',fontweight='bold') #fontfamily='serif'
    
#m.drawrivers()
m.drawstates(color='blue')
m.drawcoastlines(color='blue')
m.drawcountries(color='blue')
#m.fillcontinents(lake_color='blue')
#m.drawmapboundary(fill_color='black')
m.etopo(alpha=0.6)

m.drawparallels(np.arange(lat_0,lat_1,5),labels=[1,0,0,0],fontsize=18)
m.drawmeridians(np.arange(lon_0,lon_1,10),labels=[0,0,0,1],fontsize=18)

hgt_contour = m.contour(x, y, hgt130_500mb_smooth,hgt130_500mb_levels,colors='k',shading='flat',linewidths=3)
hgt_contourSurf = m.contour(x, y, hgt130_Surf_smooth,hgt130_Surf_levels,colors='blue',shading='flat',
                            linewidths=3,linestyle='--')

labels = '500mb'
labels2 = '1000mb'

hgt_contour.collections[0].set_label(labels)
hgt_contourSurf.collections[0].set_label(labels2)

legend = plt.legend(bbox_to_anchor=(.68, -0.05), ncol=2, prop={'size': 25},shadow=True,framealpha=1,
                        handletextpad=0.3,markerscale=8,borderpad=.5)
frame = legend.get_frame()
legend.get_frame().set_linewidth(5)
legend.get_frame().set_edgecolor("black")
#plt.clabel(hgt_contour, inline=True, fmt='%1i', fontsize=18,colors='k')

plt.title('Geopotential Heights\n'+mydate+"Z",fontsize=30)

#plt.savefig("/Users/chowdahead/Desktop/Dynamics/Intro.png")
fig

In [ ]:
plt.clf()
fig = plt.figure()

lat_0 = 20
lat_1 = 60
lon_0 = 225
lon_1 = 300   

m=Basemap(projection='mill',lat_ts=10,llcrnrlon=lon_0, \
      urcrnrlon=lon_1,llcrnrlat=lat_0,urcrnrlat=lat_1, \
      resolution='l')
    
    
#m.drawrivers()
m.drawstates()
m.drawcoastlines()
m.drawcountries()
#m.fillcontinents(lake_color='blue')
#m.drawmapboundary(fill_color='black')

m.drawparallels(np.arange(lat_0,lat_1,5),labels=[1,0,0,0],fontsize=18)
m.drawmeridians(np.arange(lon_0,lon_1,10),labels=[0,0,0,1],fontsize=18)

x, y = m(*np.meshgrid(Lon,Lat))
lons2 = [Lon[102]]
lats2 = [Lat[20]]

# Different levels to plot the contours. 
#Denhgt130_levels = np.arange(Denhgt130.min(),Denhgt130.max()+80,60)

#Smooth and re-plot the height field
#Denhgt130_smooth = ndimage.gaussian_filter(Denhgt130, sigma=1, order=0)
hgt_contour_smooth = m.contour(x, y, hgt130_500mb_smooth,hgt130_500mb_levels,colors='k',shading='flat')

hgt_contourfill=m.contourf(x,y,hgt130_500mb_smooth,hgt130_500mb_levels,shading='flat',cmap=plt.cm.jet)

# Colorbar
cb = plt.colorbar(hgt_contourfill, shrink=0.53)

# Contour labels
plt.clabel(hgt_contour_smooth, inline=True, fmt='%1i', fontsize=18,colors='k')

cb.ax.tick_params(labelsize=14)
cb.ax.set_title('Heights: $m$', fontsize=15,horizontalalignment='center',y=1.01,x=1)


# Plot title
plt.title(str(int(mylevel))+'mb Heights\n'+mydate+"Z",fontsize=30)

# Save figure, need full path to where you want the download, unless you want it in the 
# current working drive...
#plt.savefig("/Users/chowdahead/Desktop/Dynamics/500mb_Height_ContourFilled.png")

# Show the figure
fig

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h2><center><font face="fantasy, serif" color=#533563 style="font-size:40px">Functions!</font></center><br/><font face="fantasy, serif" color=#533563 style="font-size:25px">Now time to define a fucntion so we can call it many times and get all the values we need like a programmable calculator. PGF will be the name of the function and and the arguments in the parenthesis will be the variables it needs to run the calculation.</font><br/ ><h3>$PGF = -(\frac{\partial \Phi_{i}}{\partial x} \hat{i} + \frac{\partial \Phi_{j}}{\partial y} \hat{j} )= -(\frac{\Delta\Phi_{lon}}{\Delta x} \hat{i} + \frac{\Delta\Phi_{lat}}{\Delta y} \hat{j} )$<br/><br/>

$PGF = -(\frac{\Phi_{east} - \Phi_{west}}{dx} \hat{i} + \frac{\Phi_{north} - \Phi_{south}}{dy} \hat{j} ) = -g(\frac{Z_{east} - Z_{west}}{dx} \hat{i} + \frac{Z_{north} - Z_{south}}{dy} \hat{j} )$<br><br>$\left\lvert{PGF}\right\rvert = \sqrt{PGF_{x}^{2}+PGF_{y}^{2}}$</h3></h2>

In [17]:
def PGF(time,level,east,west,north,south,dx,dy):
    pgfx = -9.8*((east-west)/(dx*2))
    pgfy = -9.8*((north-south)/(dy*2))
    pgf = np.sqrt((pgfx**2)+(pgfy**2))
    return float(pgfx), float(pgfy), float(pgf)

<h2><font face="fantasy, serif" color=#533563 style="font-size:23px">Since we know the beam angle of the instrument and the lat and lon values, the dx and dy can be calcluated knowing that 2.5 deg of latitude is 278km and 2.5 deg of longitude is 213km:</font></h2>

In [18]:
dx = 213000
dy = 278000

<h2><font face="fantasy, serif" color=#533563 style="font-size:23px">Now we can calculate the geopotential heights at and around Denver. Taking the heights around Denver will allow us to calculate the PGF, Geostrophic Wind, and Vorticity</font></h2>

In [19]:
height_Denver = height[mytime][5][Denlat][Denlon]
print "Denver's height: ",height_Denver,"m"

# The reason why the northern height is Denlat-1 instead of +1 is because how the lat data was read in
height_north = height[mytime][5][Denlat-1][Denlon]
print "Height north: ",height_north,"m"

height_south = height[mytime][5][Denlat+1][Denlon]
print "Height south: ",height_south,"m"

height_east = height[mytime][5][Denlat][Denlon+1]
print "Height east: ",height_east,"m"

height_west = height[mytime][5][Denlat][Denlon-1]
print "Height west: ",height_west,"m"

Denver's height:  5362.0 m
Height north:  5390.0 m
Height south:  5362.0 m
Height east:  5339.0 m
Height west:  5392.0 m


In [20]:
pgfx,pgfy,pgf = PGF(mytime,5,height_east,height_west,height_north,height_south,dx,dy)
#print pgfx,pgfy,pgf

# We can print it out a little neater in scientific notation: (%.3e means 3 decimal places in 'e' notation)
print 'PGFx at Denver: ','%.3e' % pgfx
print 'PGFy at Denver: ','%.3e' % pgfy
print 'PGF resultant: ','%.3e' % pgf

PGFx at Denver:  1.219e-03
PGFy at Denver:  -4.935e-04
PGF resultant:  1.315e-03


In [21]:
angle = np.arccos(pgfx/pgf)*(180/np.pi)
# We can round the decimal to two places:
angle = np.around(angle,2)
print angle

22.04


<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">End of Day 1 Data, I'm Proud of You!</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:50px">Day 2 Data</font></center></h3>

<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">We can also define another function for the Coriolis factor <i>CorFac</i> with only one variable, the latitude: </font><br/><br/>$f = 2\Omega sin{\phi} \hspace{10 mm} \Omega=7.292e^{-5}s^{-1}$<br/ ></h2>

In [22]:
def CorFac(lat):
    cof = 2*7.292E-5*np.sin(Lat[lat]*np.pi/180)
    return float(cof)

In [23]:
f = CorFac(Denlat)
print 'Coriolis factor over Denver: ',f
print 'Coriolis factor over Denver: %.3e' % f

Coriolis factor over Denver:  9.37441449967e-05
Coriolis factor over Denver: 9.374e-05


<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">How would we want to scale this up to collect the Coriolis factor for all latitudes?</font></h2>

In [1]:
# Initate an empty 1d array to populate with Coriolis factors for each lat
COR = xr.DataArray(np.zeros((73)),dims=['x'],coords={'x': Lat})

for i in range(0,74):
    COR[i]=CorFac(i)
        
print "done."
print COR.shape

NameError: name 'xr' is not defined

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">End of Day 2 Data, No sweat</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:50px">Day 3 Data</font></center></h3>

<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">With the geopotential height differences and the Coriolis factor, the geostrophic winds can be calculated.</font><br/><br/><center><font style="font-size:30px">$\vec{V_{g}} = -\frac{1}{f}(\frac{\partial \Phi_{j}}{\partial y} \hat{i} - \frac{\partial \Phi_{i}}{\partial x} \hat{j} )$</font></center><br/></h2>

<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">Let's break the total horizontal geostrophic wind vector into it's zonal and meridonal components:</font></h2>

<h1>$\vec{u_{g}} = -\frac{1}{f}(\frac{\partial \Phi_{j}}{\partial y})\hspace{.1cm} \hat{i}= -\frac{g}{f}(\frac{\partial Z_{j}}{\partial y})\hspace{.1cm}\hat{i}$</h1>

In [25]:
def uGeoWind(level,time,lat,lon,dy):
    ug = (-9.8/CorFac(lat))*\
          ((height[time][level][lat-1][lon]-height[time][level][lat+1][lon])/(dy*2))   

    return ug

<h1>$\vec{v_{g}} = \frac{1}{f}(\frac{\partial \Phi_{i}}{\partial x})\hspace{.1cm}\hat{j} = \frac{g}{f}(\frac{\partial Z_{i}}{\partial x})\hspace{.1cm}\hat{j}$</h1>

In [26]:
def vGeoWind(level,time,lat,lon,dx):
    vg = (9.8/CorFac(lat))*\
          ((height[time][level][lat][lon+1]-height[time][level][lat][lon-1])/(dx*2))   
    
    return vg 

<h1>$\left\lvert{V_{g}}\right\rvert= \sqrt{u_{g}^{2}+v_{g}^{2}}$<br><br><font face="fantasy, serif" color=#533563 style="font-size:25px">We'll define another function called <i>GeoWind</i></font></h1>

In [27]:
def GeoWind(level,time,lat,lon,dx,dy):

    ug = uGeoWind(level,time,lat,lon,dy)
    vg = vGeoWind(level,time,lat,lon,dx)
    
    return np.sqrt((ug**2)+(vg**2))

In [28]:
u = uGeoWind(5,130,Denlat,Denlon,dy)
print u
v = vGeoWind(5,130,Denlat,Denlon,dx)
print v
V = GeoWind(5,130,Denlat,Denlon,dx,dy)
print V

-5.26459737697
-13.0061330906
14.0312324303


In [29]:
angle = np.arccos(u/V)*(180/np.pi)
# We can round the decimal to two places:
angle = np.around(angle,2)
print angle

print "Angle of resultant vector from x-axis:",angle,"deg"
print "In meteorlogical terms it would be:",angle+90,"deg from north \n"
print "However this is the angle the vector is pointing.","\n",\
"We're looking for the angle of origin, where the wind is coming from:",angle+90,"deg"
final_angle = angle+90
print final_angle

112.04
Angle of resultant vector from x-axis: 112.04 deg
In meteorlogical terms it would be: 202.04 deg from north 

However this is the angle the vector is pointing. 
We're looking for the angle of origin, where the wind is coming from: 202.04 deg
202.04


<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h2><center><font face="fantasy, serif" color=#533563 style="font-size:40px">Building Arrays!</font></center><br/><font face="fantasy, serif" color=#533563 style="font-size:25px">Arrays are very important and useful in data handeling. We are going to make a new array that has 2 dimensions: latitude and longitude.  Thus the size of the array is 73 for the first dimension (lats) and 144 for the second dimension (lons).<font face="fantasy, serif" color=#533563 style="font-size:25px"><br/ ></font>we can set dimensions for plotting as x and y. Notice that the dimensions seem switched with x and latitudes, but python plots with the first value on the y-axis. Instead of plotting traditionally f(x,y), python plots f(y,x).</font></h2>

<h2><font face="fantasy, serif" color=#533563 style="font-size:40px"><center>Creating new arrays for our calculations</center><br></font><font face="fantasy, serif" color=#533563 style="font-size:23px">Don't focus on the how so much as to what you are getting in the end. We are making a 2d array named WindsFull_130_500mb with coordiantes of lat and lon only. This array is focused on time 130 and level 500mb.<br><br>In the nested for-loop we will need to loop over all the lats and lons seperately so we will have a wind value for every lat/lon pair.</font></h2>

In [56]:
#%%time
WindsFull_130_500mb = xr.DataArray(np.zeros((73, 144)),dims=['x','y'],coords={'x': Lat,'y': Lon})
uWindsFull_130_500mb = xr.DataArray(np.zeros((73, 144)),dims=['x','y'],coords={'x': Lat,'y': Lon})
vWindsFull_130_500mb = xr.DataArray(np.zeros((73, 144)),dims=['x','y'],coords={'x': Lat,'y': Lon})
print WindsFull_130_500mb.shape

# We need to go from the second entry to the second to last entry, why?
for i in range(1,72):
    for j in range(1,143):
            try:
                WindsFull_130_500mb[i,j]=GeoWind(5,130,i,j,dx,dy) # What arguments does GeoWind take?
                uWindsFull_130_500mb[i,j]=uGeoWind(5,130,i,j,dy)
                vWindsFull_130_500mb[i,j]=vGeoWind(5,130,i,j,dx)
            except ZeroDivisionError:
                normalized_score = 0
                
print "done.\n"
#display(WindsFull_130_500mb)

(73, 144)
done.



In [ ]:
print "data of lon index 0 and lat index 0: ",WindsFull_130_500mb[0][0],"\n"
print "data of lon index 143 and lat index 72: ",WindsFull_130_500mb[72][143]

In [ ]:
print "Winds min value: ",WindsFull_130_500mb.min().values
print "Winds max value: ",WindsFull_130_500mb.max().values

In [42]:
%%time
WindsFull_129 = xr.DataArray(np.zeros((73, 144, 17)),dims=['x','y','z'],coords={'x': Lat,'y': Lon,'z': Level})
WindsFull_130 = xr.DataArray(np.zeros((73, 144, 17)),dims=['x','y','z'],coords={'x': Lat,'y': Lon,'z': Level})
WindsFull_131 = xr.DataArray(np.zeros((73, 144, 17)),dims=['x','y','z'],coords={'x': Lat,'y': Lon,'z': Level})

print WindsFull_130.shape

for i in range(1,72):
    for j in range(1,143):
        for k in range(0,17):
            try:
                WindsFull_129[i,j,k]=GeoWind(k,129,i,j,dx,dy)
                WindsFull_130[i,j,k]=GeoWind(k,130,i,j,dx,dy)
                WindsFull_131[i,j,k]=GeoWind(k,131,i,j,dx,dy)
            except ZeroDivisionError:
                normalized_score = WindsFull_130[i-1,j-1,k] 
print "done."
#display(WindsFull_130)

(73, 144, 17)
done.


NameError: name 'display' is not defined

In [ ]:
plt.clf()
fig = plt.figure()

lat_0 = 20
lat_1 = 60
lon_0 = 225
lon_1 = 300   

#lat_0 = -90
#lat_1 = 90
#lon_0 = 0
#lon_1 = 357.5  

m=Basemap(projection='mill',lat_ts=10,llcrnrlon=lon_0, \
      urcrnrlon=lon_1,llcrnrlat=lat_0,urcrnrlat=lat_1, \
      resolution='l')
    
    
#m.drawrivers()
m.drawstates()
m.drawcoastlines()
m.drawcountries()
#m.fillcontinents(lake_color='blue')
#m.drawmapboundary(fill_color='black')

#######################################################################
m.drawparallels(np.arange(lat_0,lat_1,5),labels=[1,0,0,0],fontsize=18)
m.drawmeridians(np.arange(lon_0,lon_1,10),labels=[0,0,0,1],fontsize=18)
#######################################################################

################################
x, y = m(*np.meshgrid(Lon,Lat))
################################

###############################
wind_levels = np.arange(0,80,1)
#wind_levels = np.arange(int(Winds.min()),int(Winds.max()),1)
###############################

#####################################################################
#Smooth and re-plot the wind field
wind_smooth = ndimage.gaussian_filter(Winds_130_500[:,:], sigma=1, order=0)
#####################################################################

###################################################################################################################
wind_contourfill = m.contourf(x, y,wind_smooth,wind_levels,cmap=plt.cm.inferno,shading='flat',antialiased = True)
hght_contour = m.contour(x, y, hgt130_500mb_smooth,hgt130_500mb_levels,colors='w',shading='flat',alpha=0.7)
###################################################################################################################

#######################################
for c in wind_contourfill.collections:
    c.set_edgecolor("face")
#######################################

###################################################################################################################
cb = plt.colorbar(wind_contourfill,alpha=0.4,orientation="vertical",shrink=0.53)
cb.ax.set_yticklabels(wind_levels[0:-1:5])
cb.ax.tick_params(labelsize=14)
cb.ax.set_title('Geo Winds: $ms^{-1}$', fontsize=15,horizontalalignment='center',y=1.01,x=1)
###################################################################################################################

plt.clabel(hght_contour, inline=True, fmt='%1i', fontsize=18,colors='w')


x2,y2 = m(*np.meshgrid(Lon[102],Lat[20]))
labels = ['Denver']
for label, xpt, ypt in zip(labels, x2, y2):   #fontname='Times New Roman'
    plt.text(xpt-440000, ypt-200000, label,fontsize=17,color='w',fontweight='bold') #fontfamily='serif'
plt.barbs(x2, y2-50000, u, v,color='w',length=9,lw=3)    

plt.title('Geostrophic Winds and '+str(int(mylevel))+'mb Heights\n'+mydate+"Z",fontsize=30)

#plt.savefig("/Users/chowdahead/Desktop/Dynamics/Isotachs_and_Winds_Contours.png")
fig

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">End of Day 3 Data, Hanging In there?</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:50px">Day 4 Data</font></center></h3>

<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">Now that we have the geostrophic winds, we can take a look at vorticity using the change in winds over distance.</font><h1>$\zeta = (\frac{\partial v_{geo}}{\partial x} - \frac{\partial u_{geo}}{\partial y} )$<br/><br><font face="fantasy, serif" color=#533563 style="font-size:25px">We can define a vorticity function, <i>Vort</i></font></h1></h2>

In [31]:
def Vort(level,time,lat,lon,dx,dy):    
    vort = (vGeoWind(level,time,lat,lon+1,dx)-vGeoWind(level,time,lat,lon-1,dx)/(2*dx))-\
                            (uGeoWind(level,time,lat-1,lon,dy)-uGeoWind(level,time,lat+1,lon,dy)/(2*dy))
    return vort

<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">We will do the same thing for the vorticity array as we did for the geostrophic winds:</font></h2>

In [33]:
Vort_130_500mb = xr.DataArray(np.zeros((73, 144)),dims=['x','y'],coords={'x': Lat,'y': Lon})

print Vort_130_500mb.shape
# We have to change the range for our for-loops, before we could use 
for i in range(1,72):
    for j in range(1,142):
        try:
            Vort_130_500mb[i,j]=Vort(5,mytime,i,j,dx,dy)
        except ZeroDivisionError:
            normalized_score = 0
print "done."


(73, 144)
done.


In [ ]:
print "Maximum vorticity value in array: ",VortFull_130_500mb.min().values
print "Minimum vorticity value in array: ",VortFull_130_500mb.max().values

In [43]:
VortFull_129 = xr.DataArray(np.zeros((73, 144, 17)),dims=['x','y','z'],coords={'x': Lat,'y': Lon, 'z': Level})
VortFull_130 = xr.DataArray(np.zeros((73, 144, 17)),dims=['x','y','z'],coords={'x': Lat,'y': Lon, 'z': Level})
VortFull_131 = xr.DataArray(np.zeros((73, 144, 17)),dims=['x','y','z'],coords={'x': Lat,'y': Lon, 'z': Level})
print VortFull_130.shape

for i in range(1,72):
    for j in range(1,143):
        for k in range(0,17):
            try:
                VortFull_129[i,j,k]=Vort(k,129,i,j,dx,dy)
                VortFull_130[i,j,k]=Vort(k,130,i,j,dx,dy)
                VortFull_131[i,j,k]=Vort(k,131,i,j,dx,dy)
            except ZeroDivisionError:
                normalized_score = VortFull_130[i-1,j-1,k] 
print "done."

(73, 144, 17)
done.


<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">We can print out info to see if the data makes sense, or we want the value at specific lats/lons, etc.</font></h2>

In [ ]:
#print "data of lon index 0 and lat index 0: "+str(Vortdata[0][0])+"\n"
#print "data of lon index 143 and lat index 72: "+str(Vortdata[72][143])

print "data of lon index 0 and lat index 0: ",VortFull_130_500mb[1][1],"\n"
print "data of lon index 143 and lat index 72: ",VortFull_130_500mb[71][142]

In [ ]:
#fig = plt.figure()
plt.clf()
lat_0 = 20
lat_1 = 60
lon_0 = 225
lon_1 = 300   

m=Basemap(projection='mill',lat_ts=10,llcrnrlon=lon_0, \
      urcrnrlon=lon_1,llcrnrlat=lat_0,urcrnrlat=lat_1, \
      resolution='l')
    
#m.drawrivers()
m.drawstates()
m.drawcoastlines()
m.drawcountries()
#m.fillcontinents(lake_color='blue')
#m.drawmapboundary(fill_color='black')
#m.etopo(alpha=0.4)
m.drawparallels(np.arange(lat_0,lat_1,5),labels=[1,0,0,0],fontsize=18)
m.drawmeridians(np.arange(lon_0,lon_1,10),labels=[0,0,0,1],fontsize=18)

x, y = m(*np.meshgrid(Lon,Lat))
vort_levels = np.arange(Vorts_130_500[:,:].min(),Vorts_130_500[:,:].max(),1)

#Smooth and re-plot the vorticity field
vort_smooth = ndimage.gaussian_filter(Vorts_130_500[:,:], sigma=1, order=0)

vort_contourfill = m.contourf(x, y, vort_smooth,vort_levels,shading='flat',cmap=plt.cm.cubehelix)
hght_contour = m.contour(x, y, hgt130_500mb_smooth,hgt130_500mb_levels,colors='w')

plt.clabel(hght_contour, inline=True, fmt='%1i', fontsize=18,colors='w')
cb = plt.colorbar(vort_contourfill, shrink=0.53)
cb.ax.tick_params(labelsize=14)
cb.ax.set_title('Vorticity: $s^{-1}$', fontsize=15,horizontalalignment='center',y=1.01,x=1)

plt.title('Vorticity and '+str(int(mylevel))+'mb Heights\n'+mydate+"Z",fontsize=30)
#plt.savefig("/Users/chowdahead/Desktop/Dynamics/VortSmooth_and_500mb_Heights.png")
fig

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">End of Day 4 Data, All Done.</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">Let's make our own NetCDF file so we can keep track of our progress:</font></center></h3>

In [57]:
dataset = Dataset('/Users/chowdahead/Desktop/Dynamics/PlottingData.nc','w')

<h2><font face="fantasy, serif" color=#533563 style="font-size:25px">We can now create variables with dimensions and keys for our new NetCDF file which we will use to avoid doing calulations everytime when we plot.</font></h2>

In [58]:
level = dataset.createDimension('level', len(Level))
lat = dataset.createDimension('lat', len(Lat))
lon = dataset.createDimension('lon', len(Lon))
date = dataset.createDimension('date', len(Time))


In [59]:
levels = dataset.createVariable('level', np.int32, ('level',))
latitudes = dataset.createVariable('latitude', np.float32,('lat',))
longitudes = dataset.createVariable('longitude', np.float32,('lon',))
dates = dataset.createVariable('dates', str, ('date',))

# Create the 3-d arrays for WInds and Vorticities
winds_129_500 = dataset.createVariable('winds_129_500', np.float32,('lat','lon',))
uwinds_129_500 = dataset.createVariable('uwinds_129_500', np.float32,('lat','lon',))
vwinds_129_500 = dataset.createVariable('vwinds_129_500', np.float32,('lat','lon',))

winds_130_500 = dataset.createVariable('winds_130_500', np.float32,('lat','lon',))
uwinds_130_500 = dataset.createVariable('uwinds_130_500', np.float32,('lat','lon',))
vwinds_130_500 = dataset.createVariable('vwinds_130_500', np.float32,('lat','lon',))

winds_131_500 = dataset.createVariable('winds_131_500', np.float32,('lat','lon',))
uwinds_131_500 = dataset.createVariable('uwinds_131_500', np.float32,('lat','lon',))
vwinds_131_500 = dataset.createVariable('vwinds_131_500', np.float32,('lat','lon',))

##########################################################################################
vorts_129_500 = dataset.createVariable('vorticity_129_500', np.float32,('lat','lon',))
vorts_130_500 = dataset.createVariable('vorticity_130_500', np.float32,('lat','lon',))
vorts_131_500 = dataset.createVariable('vorticity_131_500', np.float32,('lat','lon',))
##########################################################################################

##########################################################################################
winds_130 = dataset.createVariable('winds_130', np.float32,('lat','lon','level'))
vorts_130 = dataset.createVariable('vorticity_130', np.float32,('lat','lon','level'))
##########################################################################################

# Create the 4-d array for geopotential heights
hgts = dataset.createVariable('hgt', np.float32,('date','level','lat','lon'))


In [60]:
Date=[]
for i in range(len(Time)):
    date1 = str(Time[i])[0:10] # Remember we can take pieces of the string
    date2 = str(Time[i])[-18:-10] # some more pieces...
    date = date1+" "+date2 # Finally combine the pieces to make a cleaner date string
    
    Date.append(date)

In [62]:
for i in range(1464): 
    dates[i]=Date[i]

In [63]:
latitudes[:] = Lat
longitudes[:] = Lon
levels[:] = Level
hgts[:] = height

winds_129[:] = WindsFull_129.values
winds_130[:] = WindsFull_130.values
winds_131[:] = WindsFull_131.values

winds_129_500[:] = WindsFull_129_500mb.values
uwinds_129_500[:] = uWindsFull_129_500mb.values
vwinds_129_500[:] = vWindsFull_129_500mb.values

winds_130_500[:] = WindsFull_130_500mb.values
uwinds_130_500[:] = uWindsFull_130_500mb.values
vwinds_130_500[:] = vWindsFull_130_500mb.values

winds_131_500[:] = WindsFull_131_500mb.values
uwinds_131_500[:] = uWindsFull_131_500mb.values
vwinds_131_500[:] = vWindsFull_131_500mb.values



vorts_129[:] = VortFull_129.values
vorts_130[:] = VortFull_130.values
vorts_131[:] = VortFull_131.values

vorts_129_500[:] = Vort_129_500mb.values
vorts_130_500[:] = Vort_130_500mb.values
vorts_131_500[:] = Vort_131_500mb.values


In [64]:
dataset.close()

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>

<h3><center><font face="fantasy, serif" color=#233513 style="font-size:35px">-----------------------------//-----------------------------</font></center></h3>